local-volume/provisioner/pkg/deleter/blockcleaner_test.go

```diff
@@ -0,0 +1,211 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deleter
+
+import (
+	"github.com/kubernetes-incubator/external-storage/local-volume/provisioner/pkg/cache"
+	"github.com/kubernetes-incubator/external-storage/local-volume/provisioner/pkg/common"
+	"github.com/kubernetes-incubator/external-storage/local-volume/provisioner/pkg/util"
+	"io/ioutil"
+	"k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/record"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+)
+
+func TestBlockCleaner_GetBlockDeleteCommand(t *testing.T) {
+
+	mountConfig := common.MountConfig{HostDir: "/mnt/disks", MountDir: "/container/dir"}
+	mountConfig.BlockConf.Cleaner = common.BlockCleanupFastRest
+	mountConfig.BlockConf.Iterations = 1
+
+	// Test fast reset
+	verifyCommandCreated(t, "/container/dir/foo", mountConfig, 1, NoErrExpected, ShellCommand,
+		"-c mkfs -F \"/container/dir/foo\" && wipefs -a \"/container/dir/foo\"")
+
+	// Test zero out
+	mountConfig.BlockConf.Cleaner = common.BlockCleanupZero
+	mountConfig.BlockConf.Iterations = 5
+	verifyCommandCreated(t, "/container/dir/foo", mountConfig, 5, DDFullErrExpected, DDCommand,
+		"if=/dev/zero of=/container/dir/foo bs=4096 status=progress")
+
+	// Test randomized writes
+	mountConfig.BlockConf.Cleaner = common.BlockCleanupRandom
+	mountConfig.BlockConf.Iterations = 3
+	verifyCommandCreated(t, "/container/dir/foo", mountConfig, 1, NoErrExpected, ShredCommand,
+		"-vzf -n 3 /container/dir/foo")
+
+	// Test custom command
+	mountConfig.BlockConf.Cleaner = common.BlockCleanupCustom
+	mountConfig.BlockConf.Iterations = 6
+	mountConfig.BlockConf.CustomExec = "/bin/hello"
+	verifyCommandCreated(t, "/container/dir/foo", mountConfig, 6, NoErrExpected, ShellCommand,
+		"-c /bin/hello /container/dir/foo")
+}
+
+func TestBlockCleaner_BasicProcessExec(t *testing.T) {
+	blkCleaner := testBlockCleanerSetup(t, &testConfig{})
+	pv := createPVSpec("testpv", "/mnt/disks", v1.VolumeReleased)
+	err := blkCleaner.asyncDeleteBlockPV(pv, 1, NoErrExpected, ShellCommand, "-c", "echo 1")
```

Comment:
Do the tests actually run the actual commands?  If so, do they have the chance to accidentally mess up the host system?

Reply:
There are couple of actual commands being run. One does a basic script that only deals in stdout. The other writes 12 bytes to a temp file (which is less than most logging), but I think the code can be modified to give feedback via stdout. Let me adjust that code.

Reply:
If we do execute actual destructive commands here in the test, I'd prefer the happen in /tmp and with files/dirs that we create with an associated UUID.

I think that testHostDir is presently /mnt/disks. https://github.com/kubernetes-incubator/external-storage/blob/master/local-volume/provisioner/pkg/deleter/deleter_test.go#L32

Reply:
With the up coming change, none of the unit test commands will do any write changes. The most they do is invoke echo to stdout or exit with error code.